src/compiler/scala/tools/nsc/typechecker/Implicits.scala

```diff
@@ -529,25 +529,24 @@ trait Implicits {
       deriveTypeWithWildcards(syms.distinct)(tp)
     }
 
-    private def allSymbols(tp: Type): Set[Symbol] = {
-      @tailrec
-      def loop(tps: List[Type], acc: Set[Symbol]): Set[Symbol] = tps match {
-        case Nil => acc
-        case hd :: tl =>
-          def hdSym(syms: Set[Symbol]) = {
-            val sym = hd.typeSymbol
-            if(sym != NoSymbol) syms + sym else syms
-          }
-          hd.dealias match {
-            case SingleType(pre, sym)    => loop(pre :: hd.dealiasWiden :: tl, hdSym(acc + sym))
-            case ThisType(sym)           => loop(tl, hdSym(acc + sym))
-            case TypeRef(pre, sym, args) => loop(pre :: args ++ tl, hdSym(acc + sym))
-            case RefinedType(parents, _) => loop(parents ++ tl, hdSym(acc))
-            case _ => loop(tl, hdSym(acc))
-          }
-        }
-
-      loop(List(tp), Set.empty)
+    private object AllSymbols extends TypeCollector(Set[Symbol](NoSymbol)) {
+      def apply(tp: Type): Unit = tp match {
```

Comment:
The default implementation of a set are usually memory heavy. In the code below, the only thing we do with the set of symbols generated from this collector is to compare if they are equal. However, all the information we need for that are the identifiers of  those symbols. 

Perhaps we could use two collectors here:

- First collector builds an array of _longs_, which are the identifiers of symbols in the first type. At the end that array is sorted and repetitions removed.
- Second collector receives the array of ids (sorted and without repetitions), and uses an internal bitset to mark positions of the array whose elements it has found in the second type. It also uses a boolean variable to flag if it ever finds a symbol not in the array.

This would reduce allocations to the first array (and regrowing attempts), and the bitset.

Reply:
That sounds rather complicated - how about just using a `BitSet`?

Reply:
Trouble with a BitSet is that its size would be proportional to the _maximum_ number you store in it. So, if you are storing Ids fro, 1043 to 1258, it would take `ceiling(1258/64)` = 20 longs.

Reply:
Regardless, since Seth just scheduled this for release 2.13.3, let's go ahead with it. We can look at it later.

Reply:
Yeah sounds good for a follow up PR